docs/source/api/migrating-from-v1.rst

```diff
@@ -0,0 +1,671 @@
+..
+..
+.. Licensed under the Apache License, Version 2.0 (the "License");
+.. you may not use this file except in compliance with the License.
+.. You may obtain a copy of the License at
+..
+..     http://www.apache.org/licenses/LICENSE-2.0
+..
+.. Unless required by applicable law or agreed to in writing, software
+.. distributed under the License is distributed on an "AS IS" BASIS,
+.. WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+.. See the License for the specific language governing permissions and
+.. limitations under the License.
+..
+
+.. _to-migrating:
+
+*********************
+Migrating from API V1
+*********************
+
+Versions 1.0, 1.1, 1.2, 1.3, 1.4, and 1.5 of the :ref:`to-api` have been deprecated since Apache Traffic Control version 4.0.0 and will be removed in Apache Traffic Control version 6.0.0. Any remaining users of 1.x endpoints of the Traffic Ops API are advised to upgrade to API version 3.0.
+
+Upgrading your Traffic Ops client
+=================================
+
+The easiest way to switch to a new API version is to use the latest version of a Traffic Ops client.
+
+Golang
+------
+
+If you are using the **github.com/apache/trafficcontrol/traffic_ops/client** package, use the |to-client-library-godoc|_ package instead.
+
+.. |to-client-library-godoc| replace:: **github.com/apache/trafficcontrol/traffic_ops/v3-client**
+.. _to-client-library-godoc: https://godoc.org/github.com/apache/trafficcontrol/traffic_ops/v3-client
+
+Python
+______
+
+Update to a newer version of the `Apache-TrafficControl <https://pypi.org/project/Apache-TrafficControl>`_ package.
```

Comment:
Much as I would love to put that in, I'm not totally sure it's allowed. That's an unofficial hosting unaffiliated with our project or the ASF, so the official installation method for the Python client is still to just download the source from the repository. Maybe for the 5.0.0 release I'll add a `.whl` to the release binaries on GitHub. Food for thought.

Reply:
So we're allowed to publish to PyPI but not allowed to talk about it? That makes 0 sense, and from a cursory search, I have found 7 Apache projects that recommend installing their Python module from PyPI on their main README:

* [apache/airflow](/apache/airflow)
* [apache/dubbo-python](/apache/dubbo-python)
* [apache/dubbo-python2](/apache/dubbo-python2)
* [apache/ignite-python-thin-client](/apache/ignite-python-thin-client)
* [apache/predictionio-sdk-python](/apache/predictionio-sdk-python)
* [apache/rocketmq-client-python](/apache/rocketmq-client-python)
* [apache/skywalking-python](/apache/skywalking-python)

If there's a concrete Apache page that explicitly says *Don't tell users to download from PyPI*, that's one thing. Otherwise, I'd rather ask for forgiveness since several other Apache projects already tell users to download their module from PyPI.

Reply:
Those other projects are hosted on PyPi officially, by the ASF and published as part of the release process. That PyPi package is hosted only by me and is not updated unless and until I remember to do so. So it doesn't meet the [Release Publication Standards](https://www.apache.org/legal/release-policy.html#publication) which state:

> **Projects MUST direct outsiders towards official releases rather than raw source repositories, nightly builds, snapshots, release candidates, or any other similar packages.**

and

> As a convenience to users that might not have the appropriate tools to build a compiled version of the source, binary/bytecode packages MAY be distributed alongside official Apache releases. In all such cases, the binary/bytecode package MUST have the same version number as the source release and MUST only add binary/bytecode files that are the result of compiling that version of the source code release and its dependencies.

The PyPi package is

- not part of an official release
- using a different version number than the release to which it belongs
- not even technically the result of compiling that version of the source code, since I had to edit the README to get it to render properly on PyPi.

Also putting it on PyPi has been discussed somewhat on [the mailing list](https://lists.apache.org/thread.html/9d5b3dcb9193c44d6389a6aa580b042ec46c2896865eaadbdca8d1a2%40%3Cdev.trafficcontrol.apache.org%3E) where it seemed there wasn't widespread interest in an official hosting.

I really hope I'm being pedantic and there'd be no legal issue putting that in, but officially the way you install the client is from the source repository.

Reply:
Thanks for complete response.

> Also putting it on PyPi has been discussed somewhat on [the mailing list](https://lists.apache.org/thread.html/9d5b3dcb9193c44d6389a6aa580b042ec46c2896865eaadbdca8d1a2%40%3Cdev.trafficcontrol.apache.org%3E) where it seemed there wasn't widespread interest in an official hosting.

Wow ok. Yeah, let's take this one to the mailing list. Removed the link to PyPI in the meantime in 2942c11c84.

Reply:
> Yeah, let's take this one to the mailing list.

I'd love that. It's never been the primary topic of discussion, just tangentially related to what was going on.